torch/csrc/jit/tensorexpr/codegen.cpp

```diff
@@ -83,21 +83,134 @@ void CodeGen::call_with_numel(void** args, int64_t numel) {
       false, "This codegen backend does not implement call_with_numel");
 }
 
-StmtPtr insertAllocFree(std::unordered_set<BufPtr>& interm_bufs, StmtPtr stmt) {
+c10::optional<size_t> bufSize(BufPtr buf) {
+  size_t size = elementSize(buf->dtype().scalar_type()) * buf->dtype().lanes();
+  for (auto& d : buf->dims()) {
+    if (!d->isConstant()) {
+      return c10::nullopt;
+    }
+    size = size * (*intValue(d));
+  }
+  return size;
+}
+
+using BufRangeInfo =
+    std::unordered_map<BufPtr, std::pair<BufAccessNode, BufAccessNode>>;
+
+std::vector<std::pair<BufPtr, BufPtr>> linearScan(
+    std::unordered_set<BufPtr>& bufs,
+    BufRangeInfo& buf_ranges) {
+  // Sort buffers by the time they appear.
+  std::vector<BufPtr> bufs_sorted(bufs.begin(), bufs.end());
+  auto sorting_function_by_start_time = [&buf_ranges](
+                                            BufPtr b1, BufPtr b2) -> bool {
+    auto start1 = buf_ranges.at(b1).first;
+    auto start2 = buf_ranges.at(b2).first;
+    return std::get<2>(start1).at(0) < std::get<2>(start2).at(0);
+  };
+  std::sort(
+      bufs_sorted.begin(), bufs_sorted.end(), sorting_function_by_start_time);
+
+  auto sorting_function_by_end_time = [&buf_ranges](
+                                          BufPtr b1, BufPtr b2) -> bool {
+    auto end1 = buf_ranges.at(b1).second;
+    auto end2 = buf_ranges.at(b2).second;
+    return std::get<2>(end1).at(0) < std::get<2>(end2).at(0);
+  };
+
+  // Map intermediate buffers to the most recent used memory if any.
+  std::unordered_set<BufPtr> mm;
+  std::list<BufPtr> mm_free;
+  std::unordered_map<BufPtr, BufPtr> b2m;
+  std::vector<std::pair<BufPtr, BufPtr>> b2m_ret;
+
+  std::vector<BufPtr> buf_to_release;
+  for (auto buf : bufs_sorted) {
+    auto start = buf_ranges.at(buf).first;
+    auto end = buf_ranges.at(buf).second;
+
+    // Release memory for buffers whose live range ends before the creation time
+    // of this buf.
+    // TODO: optimize in-place opererations and copy operations
+    buf_to_release.clear();
+    for (auto& mapped : b2m) {
+      auto buf_mapped = mapped.first;
```

Comment:
Nit: maybe it's because of the var names, but it's hard for me to follow this code. I don't understand what `mapped` holds, what's in `buf_mapped`, etc. Would it be possible to use more descriptive names and add some comments to make the code more readable?